otwcode/otwarchive
# AO3-6051 AO3-4551 AO3-4829 Use MemCacheStore and overhaul skin caching

## Issue

https://otwarchive.atlassian.net/browse/AO3-6051
https://otwarchive.atlassian.net/browse/AO3-4551
https://otwarchive.atlassian.net/browse/AO3-4829

## Purpose

- Switch to using `MemCacheStore` instead of `DalliStore`. They both use `Dalli::Client` under the hood, but `MemCacheStore` makes `race_condition_ttl` work properly.
- Overhaul skin caching. The switch to `MemCacheStore` means that `Marshal.dump` is called outside of [this block in Dalli::Server](https://github.com/petergoldstein/dalli/blob/ca8b286de230546c3bfe2789e9b6bc5a2c619570/lib/dalli/server.rb#L71-L75), which means that marshaling errors will cause an error instead of silently failing and writing an error to the log. So I rewrote the caching code to cache only `skin.get_style`, and not the skin itself.

## Testing Instructions

1. As a user, make a custom skin that overrides the site skin, and start using it. Make sure that the skin is included in subsequent page loads. Then try editing the skin, and make sure that the changes are immediately reflected on the next page you visit.
2. As a user, make a custom skin that adds onto the site skin, and start using it. Make sure that the skin is included in subsequent page loads. Then try editing the skin, and make sure that the changes are immediately reflected on the next page you visit.
3. As an admin, go to Skins > Approved Skins and pick a new skin to use as the default. Make sure that the new skin is included in subsequent page loads, both for logged out users, and for users with a custom skin adding onto the default.
4. ~~As an admin, go to Skins > Approved Skins, click the link for the current default skin, and then edit it to add some CSS. Make sure that the changes are immediately reflected on the next page you visit, both for logged-out users, and for users with a custom skin adding onto the default.~~ **EDIT:** This won't necessarily work on staging/production because the filesystem isn't shared between different servers, so if the default skin is cached to the filesystem, the changes will only show up on the server that handled the `SkinsController#update` request, not any other servers.
5. As an admin, go to Skins > Approved Skins, check "Chooser" for one of the listed skins, and press "Update." Then check the footer to make sure that the skin you selected is listed under "Choose Skin."
6. As an admin, go back to the Approved Skins page, check "Not In Chooser" for the same skin, and press "Update." Then check the footer to make sure that the skin is no longer listed under "Choose Skin."
